SMA Get Configuration Endpoint Without Registry Flag No Longer Returns Configuration

michaelestrin opened the pull request:
Restores backwards compatibility.

Restores prior behavior of mapping service keys to service names in
configuration to get endpoint data for connecting to external services
to obtain configuration (and health, metrics, etc.).

Fixes: https://github.com/edgexfoundry/edgex-go/issues/1780

Signed-off-by: Michael Estrin <m.estrin@dell.com>

michaelestrin commented:
Rebased.

michaelestrin commented:
Rebased.

tsconn23 commented:
@michaelestrin Unfortunately, while the use case targeted by the issue is resolved, I think I've found a bug w/r/t Consul. Steps to reproduce below.

1.) Start Consul
2.) Run config-seed
3.) Start all services with the -r cmd line flag
4.) Make sure all services are health checking OK in Consul (http://localhost:8500/ui/dc1/services)
5.) GET http://localhost:48090/api/v1/config/edgex-core-data

On the first call, the following response is returned. Subsequent calls after the first return correctly.
```
{"configuration":{"edgex-core-data":"Get http://:0//api/v1/config: dial tcp :0: connect: can't assign requested address"}}
```
Up to you if you think this should be addressed via a separate issue/PR. The issue that this PR targets is resolved.

tsconn23 commented on internal/system/agent/init.go:
```diff
@@ -16,6 +16,7 @@ package agent
 
 import (
 	"context"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/config"
```
organize imports

michaelestrin replied:
Corrected.

michaelestrin commented:
Will investigate Consul issue tomorrow.

michaelestrin commented:
>Unfortunately, while the use case targeted by the issue is resolved, I think I've found a bug w/r/t Consul.

I'm unable to duplicate this issue on my local Windows development machine.  Is this behavior consistent on your end (e.g. you're able to duplicate the issue from current tip of master, the response is as you describe regardless of the time passed since the service's are started, etc.)?

On its surface, this issue seems similar/related to https://github.com/edgexfoundry/go-mod-core-contracts/issues/150.  

>Up to you if you think this should be addressed via a separate issue/PR. The issue that this PR targets is resolved.

I'm open to either option.

@tsconn23

michaelestrin commented:
Minor clean up.  Removed unused parameter from direct.NewMetrics().  Removed unsafe usage of ConfigurationStruct.Clients.  I doubt any of these changes has any effect but please retest @tsconn23.

tsconn23 commented:
> Minor clean up. Removed unused parameter from direct.NewMetrics(). Removed unsafe usage of ConfigurationStruct.Clients. I doubt any of these changes has any effect but please retest @tsconn23.

Retested. This morning the bug I reported above isn't happening, so I think we can go ahead and merge this PR. If it re-appears we'll treat it as a separate issue.
